harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,396 @@
+import abc
+import logging
+import queue
+import threading
+import time
+from datetime import datetime, timezone
+from typing import Any, Dict, List
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        # xxx: pass in train context here to use for metrics reporting API?
```

Comment:
that seems acceptable to me... you are already passing the distributed context as a dependency.  Feels morally equivalent.